doc/reference/trd-appid.md

```diff
@@ -0,0 +1,794 @@
+reserved for footers.
+
+This is the format of a Program Header:
+
+```
+0             2             4             6             8
++-------------+-------------+---------------------------+
+| Type (9)    | Length (16) | init_fn_offset            |
++-------------+-------------+---------------------------+
+| protected_size            | min_ram_size              |
++---------------------------+---------------------------+
+| binary_end_offset         | version                   |
++---------------------------+---------------------------+
+```
+
+It is represented in the Tock kernel with this Rust structure:
+
+```rust
+pub struct TbfHeaderV2Program {
+    init_fn_offset: u32,
+    protected_size: u32,
+    minimum_ram_size: u32,
+    binary_end_offset: u32,
+    version: u32,
+}
+```
+
+A TBF object MUST NOT have more than one Program Header. If a TBF
+Object has both a Program Header and a Main Header, the kernel's
+policy decides which is used. For example, older kernels that do not
+understand a Program Header may use the Main Header, while newer
+kernels may choose the Program Header.
+
+5.2 Credentials Footer
+-------------------------------
+
+To support credentials, the Tock Binary Format has a
+`TbfFooterV2Credentials` TLV. This TLV is variable length and has two
+fields, a 32-bit value specifying the `format` of the credentials and
+a variable length `data` field. The `format` field defines the format
+and size of the `data` field. Each value of the `format` field except
+`Reserved` MUST have a fixed data size and format. This is the format
+of a Credentials Footer:
+
+```
+0             2             4                           8
++-------------+-------------+---------------------------+
+| Type (128)  | Length      | format                    |
++-------------+-------------+---------------------------+
+| data                      |
++-------------+--------...--+
+```
+
+It is represented in the Tock kernel with this structure:
+
+```rust
+pub struct TbfFooterV2Credentials {
+    format: TbfFooterV2CredentialsType,
+    data: &[u8],
+}
+```
+
+Currently supported values of `format` are:
+
+```rust
+pub enum TbfFooterV2CredentialsType {
+    Reserved = 0,
+    CleartextID = 1,
+    Rsa3072Key = 2,
+    Rsa4096Key = 3,
+    Rsa3072KeyWithID = 4,
+    Rsa4096KeyWithID = 5,
+	SHA256 = 6,
+	SHA384 = 7,
+	SHA512 = 8,
+}
+```
+
+The `Reserved` type has a variable length. This credentials type is
+used to reserve space for future credentials (e.g., that will be added
+by another party in the deployment process). Because the `total_size`
+field of a TBF Base Header is covered by integrity, once the total
+size of the TBF Object has been decided it cannot be changed: if
+credentials need to be added later, space must be reserved for them.
+
+The `CleartextID` type has a data length of 8 bytes. It contains a
+64-bit number in big-endian format representing an application
+identifier.
```

Comment:
Why is this big-endian? This conflicts with the TBF documentation that says everything in TBF TLVs is little-endian.

Reply:
Generally speaking, network and storage formats (including those for cryptography) are big-endian. A PKCS1.5 signature, for example, interprets a hash value as the low-order bits of a big-endian value for encryption by a private key. So this definition is trying to be consistent with `Rsa3072Key` and `Rsa4096Key`, since their values are big endian.